src/common/task_runner.cpp

```diff
@@ -0,0 +1,114 @@
+/*
+ *    Copyright (c) 2021, The OpenThread Authors.
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * This file implements the Task Runner that executes tasks on the mainloop.
+ */
+
+#include "common/task_runner.hpp"
+
+#include <algorithm>
+
+#include <sys/eventfd.h>
+#include <unistd.h>
+
+#include "common/code_utils.hpp"
+
+namespace otbr {
+
+void TaskRunner::Post(const Task &aTask)
+{
+    PushTask(aTask);
+}
+
+void TaskRunner::UpdateFdSet(otSysMainloopContext &aMainloop)
+{
+    FD_SET(mEventFd, &aMainloop.mReadFdSet);
+    aMainloop.mMaxFd = std::max(mEventFd, aMainloop.mMaxFd);
+}
+
+void TaskRunner::Process(const otSysMainloopContext &aMainloop)
+{
+    if (FD_ISSET(mEventFd, &aMainloop.mReadFdSet))
+    {
+        PopTasks();
+    }
+}
+
+TaskRunner::TaskRunner(void)
+{
+    mEventFd = eventfd(/* initval */ 0, EFD_SEMAPHORE);
+
+    // There are no chances that we can handle failure of creating
+    // the event fd, simply die.
+    VerifyOrDie(mEventFd != -1, strerror(errno));
+}
+
+TaskRunner::~TaskRunner(void)
+{
+    if (mEventFd != -1)
+    {
+        close(mEventFd);
```

Comment:
Just in case this destructor is called multiple times.
```suggestion
        close(mEventFd);
        mEventFd = -1;
```